### DISCUSSION TITLE
Pass-through transparency (v11)

### DISCUSSION THREAD

**Discussion by mixonomicon** [AvaloniaUI/Avalonia#11911] [Q&A] [answered]
I tried to port some code from Windows WPF which shows partially transparent background (no chrome) window. In WPF, the allowed transparency allowed user to click within the transparent area and have the click be processed by the app in the background. e.g. behaving like the transparent background was not even part of the app. 

When I tried to do the same in AvaloniaUI, I did get the transparency working, on both Linux and Windows, using code below and an app.manifests which states support for Win10. But the transparent area is clearly a part of the app, on both OSes, clicking on it does nothing, and mouse hint is the one for the app, not the background (for example if you start the app in front of the editor, the mouse is an arrow until it leaves the invisible transparent section of the background, when it gets shown as an edit mouse pointer). 

Is there some way to achieve true functional transparency? 

``` xaml
<Window xmlns="bunch of namespaces..."
      ....
      x:Name="LauncherWindow" x:Class="LauncherWindow"
      Width="1200" Height="800"
      Title="MyTestApp" 
      WindowStartupLocation="CenterScreen" 
      ShowInTaskbar="False" 
      SizeToContent="WidthAndHeight" 
      Topmost="True"
      TransparencyLevelHint="Transparent"
      TransparencyBackgroundFallback="Transparent"
      Background="Transparent"
      SystemDecorations="None"
      >
    <!-- does not seem to be needed in v11
      ExtendClientAreaToDecorationsHint="True"
      ExtendClientAreaChromeHints="NoChrome"
      ExtendClientAreaTitleBarHeightHint="-1"-->
```

**Comment by kekekeks** [comment, marked as answer]
1) with X11 click-through transparency is only available with XShape extension. You can't do it with alpha-channel alone on that platform
2) On Windows we are not using `UpdateLayeredWindow` since it's not compatible with GPU rendering. You can try playing with SetWindowRgn

**Reply by cedric56** [reply]
Hello, how can i achieve click through with XShape ? Iam trying to make an overlay window. Do you have any example ? thanks

**Reply by thevortexcloud** [reply]
There is not really a lot of examples for it online, especially C# ones. It's a fairly obscure extension in the first place. There are some C examples. EG XEyes and OClock. 

https://gitlab.freedesktop.org/xorg/app/xeyes
https://gitlab.freedesktop.org/xorg/app/oclock

Otherwise it's mostly just reading the man pages for the extension and trying to work backwards.

That said, I am not sure how well/if it even works with Wayland/XWayland. Which is quickly becoming a defacto default on Linux.

**Comment by mixonomicon** [comment]
@kekekeks 

Yea, wont work here though. `SetWindowRgn` is a performance killer. The use case here was a transitioning animation, a splash screen shows up, inits, and then "blows up" in size, hides splash elements and brings main content into perspective. A story board with an interplay of resizing and opacity, in essence. 

On WPF, initially this was a failure because animating Width and Height of a window led to SetWindowRgn being animated, and that was so slow that animation was very very slow and choppy (didnt really waste too much time on it, but I believe it caused a repaint for every little change of the size). We solved it back then by just having a large transparent Window and a splash in the middle, and then we just animate the splash to full extent and then transition it to enabled background. Since Width and Height were not animated, it ran fast :) it was dirty but it worked. 

But it is what it is, Wont kill the migration initiative just because of this. We'll change the animation itself and use a large splash screen. I am more concerned by the lack of 3rd party commercial controls (and their lack of interest to make them) than by this. 

Thanks for your reply.

**Reply by Butcher-vj** [reply]
Have you achieved the transparency click through ?

**Comment by teta2k** [comment]
I found a similar scenario [here](https://github.com/glfw/glfw/issues/1512) where clicks manage to get through by setting a black color key with `SetLayeredWindowAttributes(handle, RGB(0, 0, 0), 255, LWA_COLORKEY);`

Unfortunately I couldn't make it work with Avalonia.. Am I missing something?
